ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BlockAttestationsPool.java

```diff
@@ -0,0 +1,168 @@
+import java.util.Comparator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.PriorityBlockingQueue;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.storage.ChainStorage;
+import tech.pegasys.artemis.util.SSZTypes.Bitlist;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class BlockAttestationsPool {
+
+  private final ConcurrentHashMap<Bytes32, UnsignedLong> dataRootToSlot = new ConcurrentHashMap<>();
+
+  private final ConcurrentHashMap<Bytes32, Bitlist> unprocessedAttestationsBitlist =
+      new ConcurrentHashMap<>();
+
+  private final ConcurrentHashMap<Bytes32, Bitlist> processedAttestationsBitlist =
+      new ConcurrentHashMap<>();
+
+  private final int QUEUE_MAX_SIZE =
+      Constants.MAX_VALIDATORS_PER_COMMITTEE
+          * Constants.MAX_COMMITTEES_PER_SLOT
+          * Constants.SLOTS_PER_EPOCH;
+
+  private final Queue<Attestation> aggregateAttesationsQueue =
+      new PriorityBlockingQueue<>(QUEUE_MAX_SIZE, Comparator.comparing(a -> a.getData().getSlot()));
+
+  public void addUnprocessedAttestationToQueue(Attestation newAttestation) {
+    if (!isSingleAttester(newAttestation)) {
+
+      Bytes32 attestationDataHash = newAttestation.getData().hash_tree_root();
+      ChainStorage.get(attestationDataHash, this.unprocessedAttestationsBitlist)
+          .ifPresentOrElse(
+              oldBitlist -> {
+                if (representsNewAttesterAggregate(oldBitlist, newAttestation)) {
+
+                  for (int i = 0; i < newAttestation.getAggregation_bits().getCurrentSize(); i++) {
+                    if (newAttestation.getAggregation_bits().getBit(i) == 1) oldBitlist.setBit(i);
+                  }
+
+                  ChainStorage.add(newAttestation, aggregateAttesationsQueue);
+
+                  ChainStorage.add(
+                      attestationDataHash,
+                      newAttestation.getData().getTarget().getEpoch(),
+                      dataRootToSlot);
+                }
+              },
+              () -> {
+                ChainStorage.add(newAttestation, aggregateAttesationsQueue);
+
+                ChainStorage.add(
+                    attestationDataHash,
+                    newAttestation.getAggregation_bits().copy(),
+                    this.unprocessedAttestationsBitlist);
+
+                ChainStorage.add(
+                    attestationDataHash,
+                    newAttestation.getData().getTarget().getEpoch(),
+                    dataRootToSlot);
+              });
+    }
+  }
+
+  public void addAttestationProcessedInBlock(Attestation attestation) {
+    Bytes32 attestationDataHash = attestation.getData().hash_tree_root();
+    ChainStorage.get(attestationDataHash, this.processedAttestationsBitlist)
+        .ifPresentOrElse(
```

Comment:
Similar issue here - we want to use the `ConcurrentMap` methods that allow for atomic updates.

Reply:
done.